File: jobs/metrics.py
"""Pull her results from the stats connections (BUILD_PLAN.md sections 3 step 8, 6, phase 8).

Spec (worker/jobs/metrics.ts buildSpec):
  posts       [{id, platform, url, posted_at}]  dashboard posts that went out, to match by link
  instagram   {access_token, user_id} | null    Instagram API with Instagram Login (refreshed by the Worker)
  youtube     {access_token, channel_id} | null  Google token (refreshed by the Worker)
  max_videos  int

Result: {"videos": [...], "accounts": [...], "errors": [{provider, kind}], "providers": [...],
"metrics": []}. Each video carries post_id when its link matches a dashboard post; the Worker
stores the numbers in platform_videos and metrics and recomputes the learned posting slots.
Numbers and ids only in logs (section 13).
"""
from __future__ import annotations

import json
import re
import urllib.error
import urllib.parse
import urllib.request
from datetime import date
from typing import Any

from common import Job, log, run

IG = "https://graph.instagram.com/v21.0"
YT = "https://www.googleapis.com/youtube/v3"
YTA = "https://youtubeanalytics.googleapis.com/v2/reports"


class ProviderError(Exception):
    def __init__(self, kind: str) -> None:
        super().__init__(kind)
        self.kind = kind  # expired | rate_limit | failed


def get_json(url: str, token: str | None = None) -> dict[str, Any]:
    headers = {"Authorization": f"Bearer {token}"} if token else {}
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=60) as res:
            return json.loads(res.read().decode())
    except urllib.error.HTTPError as e:
        body = e.read().decode(errors="replace")
        if e.code == 401 or '"code":190' in body.replace(" ", ""):
            raise ProviderError("expired") from e
        if e.code == 429 or "quotaExceeded" in body or "rateLimit" in body or '"code":4,' in body.replace(" ", ""):
            raise ProviderError("rate_limit") from e
        raise ProviderError("failed") from e
    except (urllib.error.URLError, TimeoutError) as e:
        raise ProviderError("failed") from e


def norm_link(url: str | None) -> str:
    if not url:
        return ""
    u = urllib.parse.urlsplit(url.strip())
    return f"{u.netloc.lower().removeprefix('www.')}{u.path.rstrip('/')}"


def yt_id(url: str | None) -> str | None:
    if not url:
        return None
    m = re.search(r"(?:shorts/|v=|youtu\.be/)([A-Za-z0-9_-]{11})", url)
    return m.group(1) if m else None


def iso8601_seconds(s: str | None) -> float | None:
    return float(s) if s not in (None, "") else None


# ---------- Instagram ----------

def instagram(cfg: dict[str, Any], posts: list[dict[str, Any]], max_videos: int) -> tuple[list[dict[str, Any]], dict[str, Any]]:
    tok = cfg["access_token"]
    q = urllib.parse.urlencode
    me = get_json(f"{IG}/me?{q({'fields': 'user_id,username,followers_count,media_count', 'access_token': tok})}")
    by_link = {norm_link(p["url"]): p["id"] for p in posts if p["platform"] == "instagram"}
    videos: list[dict[str, Any]] = []
    url = f"{IG}/me/media?{q({'fields': 'id,permalink,timestamp,media_type,media_product_type,like_count,comments_count,caption', 'limit': 50, 'access_token': tok})}"
    while url and len(videos) < max_videos:
        page = get_json(url)
        for m in page.get("data") or []:
            if m.get("media_type") != "VIDEO" and m.get("media_product_type") != "REELS":
                continue
            stats = {"views": 0, "likes": m.get("like_count") or 0, "comments": m.get("comments_count") or 0, "shares": 0, "saved": 0, "ig_reels_avg_watch_time": None}
            for metrics in ("views,likes,comments,shares,saved,ig_reels_avg_watch_time", "views,likes,comments,shares,saved", "reach,likes,comments,shares,saved"):
                try:
                    ins = get_json(f"{IG}/{m['id']}/insights?{q({'metric': metrics, 'access_token': tok})}")
                    for row in ins.get("data") or []:
                        vals = row.get("values") or [{}]
                        name = "views" if row.get("name") == "reach" and "views" not in metrics else row.get("name")
                        stats[name] = vals[0].get("value", 0)
                    break
                except ProviderError as e:
                    if e.kind != "failed":
                        raise
            watch = stats.get("ig_reels_avg_watch_time")
            videos.append({
                "platform": "instagram",
                "external_id": m["id"],
                "url": m.get("permalink"),
                "title": (m.get("caption") or "")[:120] or None,
                "posted_at": m.get("timestamp"),
                "views": int(stats.get("views") or 0),
                "likes": int(stats.get("likes") or 0),
                "comments": int(stats.get("comments") or 0),
                "shares": int(stats.get("shares") or 0),
                "saves": int(stats.get("saved") or 0),
                "avg_watch_s": (float(watch) / 1000.0) if watch else None,  # reported in milliseconds
                "post_id": by_link.get(norm_link(m.get("permalink"))),
            })
        url = (page.get("paging") or {}).get("next")
    recent = sorted(videos, key=lambda v: v["posted_at"] or "", reverse=True)[:30]
    account = {"platform": "instagram", "followers": int(me.get("followers_count") or 0), "avg_views": int(sum(v["views"] for v in recent) / len(recent)) if recent else 0}
    log("metrics.instagram", videos=len(videos), matched=sum(1 for v in videos if v["post_id"]))
    return videos, account


# ---------- YouTube ----------

def youtube(cfg: dict[str, Any], posts: list[dict[str, Any]], max_videos: int) -> tuple[list[dict[str, Any]], dict[str, Any]]:
    tok = cfg["access_token"]
    q = urllib.parse.urlencode
    ch = get_json(f"{YT}/channels?{q({'part': 'statistics,contentDetails', 'mine': 'true'})}", tok)
    item = (ch.get("items") or [{}])[0]
    uploads = ((item.get("contentDetails") or {}).get("relatedPlaylists") or {}).get("uploads")
    followers = int((item.get("statistics") or {}).get("subscriberCount") or 0)
    ids: list[str] = []
    token = None
    while uploads and len(ids) < max_videos:
        params = {"part": "contentDetails", "playlistId": uploads, "maxResults": 50}
        if token:
            params["pageToken"] = token
        page = get_json(f"{YT}/playlistItems?{q(params)}", tok)
        ids += [i["contentDetails"]["videoId"] for i in page.get("items") or []]
        token = page.get("nextPageToken")
        if not token:
            break
    ids = ids[:max_videos]
    by_id = {yt_id(p["url"]): p["id"] for p in posts if p["platform"] == "youtube" and yt_id(p["url"])}
    videos: list[dict[str, Any]] = []
    for i in range(0, len(ids), 50):
        batch = ids[i : i + 50]
        data = get_json(f"{YT}/videos?{q({'part': 'statistics,snippet', 'id': ','.join(batch)})}", tok)
        for v in data.get("items") or []:
            st = v.get("statistics") or {}
            videos.append({
                "platform": "youtube",
                "external_id": v["id"],
                "url": f"https://youtube.com/shorts/{v['id']}",
                "title": ((v.get("snippet") or {}).get("title") or "")[:120] or None,
                "posted_at": (v.get("snippet") or {}).get("publishedAt"),
                "views": int(st.get("viewCount") or 0),
                "likes": int(st.get("likeCount") or 0),
                "comments": int(st.get("commentCount") or 0),
                "shares": 0,
                "saves": 0,
                "avg_watch_s": None,
                "post_id": by_id.get(v["id"]),
            })
    # Average view duration and shares from YouTube Analytics (best effort).
    if videos:
        try:
            for i in range(0, len(videos), 100):
                chunk = videos[i : i + 100]
                rep = get_json(f"{YTA}?{q({'ids': 'channel==MINE', 'startDate': '2015-01-01', 'endDate': date.today().isoformat(), 'metrics': 'averageViewDuration,shares', 'dimensions': 'video', 'filters': 'video==' + ','.join(v['external_id'] for v in chunk), 'maxResults': 200})}", tok)
                rows = {r[0]: r for r in rep.get("rows") or []}
                for v in chunk:
                    r = rows.get(v["external_id"])
                    if r:
                        v["avg_watch_s"] = iso8601_seconds(r[1])
                        v["shares"] = int(r[2] or 0)
        except ProviderError as e:
            if e.kind == "expired":
                raise
            log("metrics.youtube.analytics_skipped", kind=e.kind)
    recent = sorted(videos, key=lambda v: v["posted_at"] or "", reverse=True)[:30]
    account = {"platform": "youtube", "followers": followers, "avg_views": int(sum(v["views"] for v in recent) / len(recent)) if recent else 0}
    log("metrics.youtube", videos=len(videos), matched=sum(1 for v in videos if v["post_id"]))
    return videos, account


def main(job: Job, spec: dict[str, Any]) -> dict[str, Any]:
    posts = spec.get("posts") or []
    max_videos = int(spec.get("max_videos") or 60)
    out: dict[str, Any] = {"videos": [], "accounts": [], "errors": [], "providers": [], "metrics": []}
    plan = [("meta", spec.get("instagram"), instagram), ("google", spec.get("youtube"), youtube)]
    for i, (provider, cfg, fn) in enumerate(plan):
        job.progress("reading stats", i, len(plan))
        if not cfg:
            continue
        try:
            videos, account = fn(cfg, posts, max_videos)
            out["videos"] += videos
            out["accounts"].append(account)
            out["providers"].append(provider)
        except ProviderError as e:
            out["errors"].append({"provider": provider, "kind": e.kind})
            log("metrics.provider", provider=provider, ok=False, kind=e.kind)
    job.progress("reading stats", len(plan), len(plan))
    log("metrics.done", videos=len(out["videos"]), errors=len(out["errors"]))
    return out


if __name__ == "__main__":
    run(main)

File: worker/jobs/metrics.ts
// Pull her results from the stats connections (BUILD_PLAN.md sections 3 step 8, 6, phase 8).
// jobs/metrics.py reads Instagram (Instagram API with Instagram Login: media + insights) and
// YouTube (Data API statistics + Analytics average view duration) with the tokens handed to it
// in the signed spec, matches her videos to dashboard posts by link, and reports numbers only.
//
// Learning loop: after every sync (and every TikTok import) `updateLearnedSlots` recomputes her
// best posting slots from real results. A platform gets learned slots only once it has at least
// 4 weeks of posted history (domain/learning.ts). They are stored in the settings key
// `learned_slots` as Partial<Record<Platform, Slot[]>> for the Calendar planner to read;
// `posting_slots_source` is left as it is (the planner owns that switch).
import type { JobHandler } from "./registry";
import type { Env } from "../env";
import { PLATFORMS, type Platform, type Slot } from "@shared/constants";
import { markConnection } from "../lib/connections";
import { recordEvent, setHealth, setSetting } from "../lib/db";
import { newId, nowIso } from "../lib/ids";
import { log } from "../lib/log";
import { learnSlots, type Observation } from "../domain/learning";
import { readSettings } from "../routes/settings";
import { freshToken, markStatsConnectionFailed, STATS_HEALTH_NAME, STATS_RECONNECT_GUIDE, type StatsProvider } from "../routes/oauth";

export interface VideoRow {
  platform: Platform;
  external_id: string;
  url: string | null;
  title: string | null;
  posted_at: string | null;
  views: number;
  likes: number;
  comments: number;
  shares: number;
  saves: number;
  avg_watch_s: number | null;
  /** Set by the job when it matched this video to a dashboard post by link. */
  post_id?: string | null;
}

interface MetricRow {
  post_id: string;
  views: number;
  likes: number;
  comments: number;
  shares: number;
  saves: number;
  avg_watch_s: number | null;
}

interface MetricsResult {
  metrics?: MetricRow[];
  videos?: VideoRow[];
  accounts?: { platform: Platform; followers: number; avg_views: number }[];
  errors?: { provider: StatsProvider; kind: "expired" | "rate_limit" | "failed" }[];
  providers?: StatsProvider[];
}

const n = (x: unknown) => Math.max(0, Math.floor(Number(x) || 0));

/** Insert or refresh her videos; links them to a dashboard post when the link matches. */
export async function upsertVideos(env: Env, videos: VideoRow[], source: "api" | "import"): Promise<number> {
  const now = nowIso();
  let count = 0;
  for (const v of videos) {
    if (!PLATFORMS.includes(v.platform) || !v.external_id) continue;
    await env.DB.prepare(
      `INSERT INTO platform_videos (id, platform, external_id, url, title, posted_at, views, likes, comments, shares, saves, avg_watch_s, source, post_id, captured_at)
       VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, COALESCE((SELECT id FROM posts WHERE id = ?15), (SELECT id FROM posts WHERE url IS NOT NULL AND url = ?4 LIMIT 1)), ?14)
       ON CONFLICT(platform, external_id) DO UPDATE SET
         url = COALESCE(excluded.url, platform_videos.url), title = COALESCE(excluded.title, platform_videos.title),
         posted_at = COALESCE(excluded.posted_at, platform_videos.posted_at), views = excluded.views, likes = excluded.likes,
         comments = excluded.comments,
         -- A source that cannot see shares / saves (YouTube's public numbers) reports 0: keep what a richer read found.
         shares = CASE WHEN excluded.shares > 0 THEN excluded.shares ELSE platform_videos.shares END,
         saves = CASE WHEN excluded.saves > 0 THEN excluded.saves ELSE platform_videos.saves END,
         avg_watch_s = COALESCE(excluded.avg_watch_s, platform_videos.avg_watch_s), source = excluded.source,
         post_id = COALESCE(excluded.post_id, platform_videos.post_id), captured_at = excluded.captured_at`,
    )
      .bind(newId("pv"), v.platform, String(v.external_id).slice(0, 300), v.url?.slice(0, 500) ?? null, v.title?.slice(0, 300) ?? null, v.posted_at, n(v.views), n(v.likes), n(v.comments), n(v.shares), n(v.saves), v.avg_watch_s ?? null, source, now, v.post_id ?? null)
      .run();
    count++;
  }
  // A dashboard post that matched a platform video gets its numbers in `metrics` too.
  await env.DB.prepare(
    `INSERT INTO metrics (id, post_id, captured_at, views, likes, comments, shares, saves, avg_watch_s)
     SELECT 'met_' || lower(hex(randomblob(6))), pv.post_id, ?, pv.views, pv.likes, pv.comments, pv.shares, pv.saves, pv.avg_watch_s
     FROM platform_videos pv WHERE pv.post_id IS NOT NULL AND pv.captured_at = ?`,
  )
    .bind(now, now)
    .run();
  return count;
}

/** Every result with a posting time: her platform videos plus dashboard posts with metrics. */
export async function loadObservations(env: Env): Promise<(Observation & { recipe: string | null })[]> {
  const { results } = await env.DB.prepare(
    `SELECT pv.platform AS platform, pv.posted_at AS posted_at, pv.views AS views,
            (SELECT c.recipe FROM posts p JOIN clips c ON c.id = p.clip_id WHERE p.id = pv.post_id) AS recipe
       FROM platform_videos pv WHERE pv.posted_at IS NOT NULL
     UNION ALL
     SELECT p.platform, COALESCE(p.posted_at, p.scheduled_at),
            (SELECT m.views FROM metrics m WHERE m.post_id = p.id ORDER BY m.captured_at DESC LIMIT 1), c.recipe
       FROM posts p JOIN clips c ON c.id = p.clip_id
      WHERE p.status = 'posted' AND EXISTS (SELECT 1 FROM metrics m WHERE m.post_id = p.id)
        AND p.id NOT IN (SELECT post_id FROM platform_videos WHERE post_id IS NOT NULL)`,
  ).all<{ platform: Platform; posted_at: string; views: number | null; recipe: string | null }>();
  return results.map((r) => ({ platform: r.platform, posted_at: r.posted_at, views: r.views ?? 0, recipe: r.recipe }));
}

/** Recompute learned slots from real results and store them for the planner. */
export async function updateLearnedSlots(env: Env): Promise<Partial<Record<Platform, Slot[]>>> {
  const s = await readSettings(env);
  const learned = learnSlots(await loadObservations(env), s.audience_timezone, s.weekly_caps);
  if (Object.keys(learned).length > 0) {
    await setSetting(env.DB, "learned_slots", learned);
    await setSetting(env.DB, "learned_slots_at", nowIso());
    await recordEvent(env.DB, "learning.slots", null, { platforms: Object.keys(learned).length });
  }
  log.info("learning.slots", { platforms: Object.keys(learned).length });
  return learned;
}

async function buildSpec(env: Env, jobId: string) {
  const { results: posts } = await env.DB.prepare(
    "SELECT id, platform, url, COALESCE(posted_at, scheduled_at) AS posted_at FROM posts WHERE status = 'posted' AND url IS NOT NULL AND COALESCE(posted_at, scheduled_at) >= ? ORDER BY posted_at DESC LIMIT 500",
  )
    .bind(new Date(Date.now() - 180 * 86_400_000).toISOString())
    .all<{ id: string; platform: Platform; url: string; posted_at: string }>();
  const meta = await freshToken(env, "meta");
  const google = await freshToken(env, "google");
  return {
    job_id: jobId,
    type: "metrics",
    posts,
    instagram: meta ? { access_token: meta.access_token, user_id: meta.account_id } : null,
    youtube: google ? { access_token: google.access_token, channel_id: google.account_id } : null,
    max_videos: 60,
  };
}

async function connectedProviders(env: Env): Promise<StatsProvider[]> {
  const { results } = await env.DB.prepare("SELECT service FROM connections WHERE service IN ('meta', 'google') AND status = 'ok'").all<{ service: StatsProvider }>();
  return results.map((r) => r.service);
}

export const metricsJob: JobHandler = {
  async buildSpec(env, jobId) {
    return buildSpec(env, jobId);
  },

  async applyResult(env, jobId, refId, raw) {
    void refId;
    const r = (raw ?? {}) as MetricsResult;
    let metricRows = 0;
    for (const m of r.metrics ?? []) {
      if (typeof m?.post_id !== "string") continue;
      const res = await env.DB.prepare(
        "INSERT INTO metrics (id, post_id, captured_at, views, likes, comments, shares, saves, avg_watch_s) SELECT ?, id, ?, ?, ?, ?, ?, ?, ? FROM posts WHERE id = ?",
      )
        .bind(newId("met"), nowIso(), n(m.views), n(m.likes), n(m.comments), n(m.shares), n(m.saves), m.avg_watch_s ?? null, m.post_id)
        .run();
      metricRows += res.meta.changes ?? 0;
    }
    const videos = await upsertVideos(env, (r.videos ?? []).filter((v) => v.platform === "instagram" || v.platform === "youtube"), "api");
    for (const a of r.accounts ?? []) {
      if (!PLATFORMS.includes(a.platform)) continue;
      await env.DB.prepare("INSERT INTO account_stats (id, platform, captured_at, followers, avg_views, source) VALUES (?, ?, ?, ?, ?, 'api')").bind(newId("acs"), a.platform, nowIso(), n(a.followers), n(a.avg_views)).run();
    }
    const errored = new Set<StatsProvider>();
    for (const e of r.errors ?? []) {
      if (e.provider !== "meta" && e.provider !== "google") continue;
      errored.add(e.provider);
      const who = e.provider === "meta" ? "Instagram" : "YouTube";
      if (e.kind === "expired") await markStatsConnectionFailed(env, e.provider, `${who} needs you to reconnect.`);
      else await setHealth(env.DB, STATS_HEALTH_NAME[e.provider], "yellow", e.kind === "rate_limit" ? `${who} asked us to slow down. We'll try again next week.` : `${who} didn't answer this time. We'll try again next week.`, STATS_RECONNECT_GUIDE[e.provider]);
    }
    const ran = r.providers ?? (await connectedProviders(env));
    for (const p of ran) {
      if (errored.has(p)) continue;
      const count = (r.videos ?? []).filter((v) => v.platform === (p === "meta" ? "instagram" : "youtube")).length;
      await markConnection(env, p, "ok", null, { last_sync_at: nowIso(), videos: count });
      await setHealth(env.DB, STATS_HEALTH_NAME[p], "green", `Synced · ${count} videos`, null);
    }
    await updateLearnedSlots(env);
    await recordEvent(env.DB, "metrics.synced", jobId, { metrics: metricRows, videos });
    log.info("metrics.apply", { metrics: metricRows, videos, errors: errored.size });
  },

  async onFailure(env, jobId, refId, safeError) {
    void refId;
    for (const p of await connectedProviders(env)) await setHealth(env.DB, STATS_HEALTH_NAME[p], "yellow", "The last stats sync did not finish. It runs again next week.", STATS_RECONNECT_GUIDE[p]);
    await recordEvent(env.DB, "metrics.failed", jobId, { len: safeError.length });
    log.warn("metrics.failed");
  },

  /**
   * Fake runner: plausible numbers for every posted dashboard post, plus six weeks of her
   * Instagram / YouTube history for each connected provider (enough for the learning loop).
   * options.fail = "meta" | "google" returns the expired-token failure shape for that provider.
   */
  async fakeRun(env, jobId, refId, options) {
    void refId;
    const spec = await buildSpec(env, jobId);
    let seed = 7;
    const rnd = () => ((seed = (seed * 16807) % 2147483647) / 2147483647);
    const metrics: MetricRow[] = spec.posts.map((p) => {
      const views = Math.round(400 + rnd() * 9000);
      return { post_id: p.id, views, likes: Math.round(views * 0.06), comments: Math.round(views * 0.008), shares: Math.round(views * 0.01), saves: Math.round(views * 0.012), avg_watch_s: Math.round(6 + rnd() * 14) };
    });
    const videos: VideoRow[] = [];
    const accounts: MetricsResult["accounts"] = [];
    const errors: MetricsResult["errors"] = [];
    const providers: StatsProvider[] = [];
    const plan: [StatsProvider, Platform, boolean][] = [
      ["meta", "instagram", !!spec.instagram],
      ["google", "youtube", !!spec.youtube],
    ];
    for (const [prov, platform, connected] of plan) {
      if (options.fail === prov) {
        errors.push({ provider: prov, kind: "expired" });
        continue;
      }
      if (!connected) continue;
      providers.push(prov);
      // 18 videos over the last 6 weeks; evenings do better, so the learning loop has a signal.
      for (let i = 0; i < 18; i++) {
        const d = new Date(Date.now() - (i * 42 * 86_400_000) / 18 - 86_400_000);
        const hour = [9, 12, 16, 19, 20, 21][i % 6];
        d.setUTCHours(hour + 4, 0, 0, 0);
        const evening = hour >= 19 ? 2.2 : 1;
        const views = Math.round((platform === "instagram" ? 1800 : 900) * evening * (0.6 + rnd()));
        videos.push({ platform, external_id: `fake_${platform}_${i}`, url: null, title: null, posted_at: d.toISOString(), views, likes: Math.round(views * 0.05), comments: Math.round(views * 0.006), shares: Math.round(views * 0.01), saves: Math.round(views * 0.01), avg_watch_s: Math.round(8 + rnd() * 12) });
      }
      const mine = videos.filter((v) => v.platform === platform);
      accounts.push({ platform, followers: platform === "instagram" ? 4820 : 1260, avg_views: Math.round(mine.reduce((s, v) => s + v.views, 0) / mine.length) });
    }
    return { metrics, videos, accounts, errors, providers };
  },
};
